src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -978,66 +979,101 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
 
-                case object val when LanguagePrimitives.TryConvertTo(val, out MatchEvaluator matchEvaluator):
-                    return regex.Replace(input, matchEvaluator);
+                    default:
+                        _cachedReplacementString = PSObject.ToStringParser(_context, substitute);
+                        break;
+                }
 
-                default:
-                    string replacement = PSObject.ToStringParser(context, substitute);
-                    return regex.Replace(input, replacement);
+                // Local helper function to avoid creating an instance of the generated delegate helper class
+                // every time 'ReplaceOperatorImpl' is invoked.
+                static MatchEvaluator GetMatchEvaluator(ExecutionContext context, ScriptBlock sb)
+                {
+                    return match =>
+                    {
+                        var result = sb.DoInvokeReturnAsIs(
+                            useLocalScope: false, /* Use current scope to be consistent with 'ForEach/Where-Object {}' and 'collection.ForEach{}/Where{}' */
+                            errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToCurrentErrorPipe,
+                            dollarUnder: match,
+                            input: AutomationNull.Value,
+                            scriptThis: AutomationNull.Value,
+                            args: Array.Empty<object>());
+
+                        return PSObject.ToStringParser(context, result);
+                    };
+                }
             }
 
-            // Local helper function to avoid creating an instance of the generated delegate helper class
-            // every time 'ReplaceOperatorImpl' is invoked.
-            static MatchEvaluator GetMatchEvaluator(ExecutionContext context, ScriptBlock sb)
+            /// <summary>
+            /// ReplaceOperator implementation.
+            /// Abstracts away conversion of the optional substitute parameter to either a string or a MatchEvaluator delegate
+            /// and finally returns the result of the final Regex.Replace operation.
+            /// </summary>
+            public object Replace(string input)
             {
-                return match =>
+                if (_cachedReplacementString is not null)
                 {
-                    var result = sb.DoInvokeReturnAsIs(
-                        useLocalScope: false, /* Use current scope to be consistent with 'ForEach/Where-Object {}' and 'collection.ForEach{}/Where{}' */
-                        errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToCurrentErrorPipe,
-                        dollarUnder: match,
-                        input: AutomationNull.Value,
-                        scriptThis: AutomationNull.Value,
-                        args: Array.Empty<object>());
+                    return _regex.Replace(input, _cachedReplacementString);
+                }
+                else
```

Comment:
Similar to what @rjmholt's comment, this else block is redundant.

Reply:
Done.